core/src/main/java/com/linecorp/armeria/internal/client/DefaultDnsNameResolver.java

```diff
@@ -103,10 +118,39 @@ public void operationComplete(Future<List<DnsRecord>> future) throws Exception {
             }
         };
 
-        questions.forEach(q -> delegate.resolveAll(q).addListener(listener));
+        final Builder<Promise<List<DnsRecord>>> promises =
+                ImmutableList.builderWithExpectedSize(questions.size());
+        questions.forEach(q -> {
+            final Promise<List<DnsRecord>> promise = executor().newPromise();
+            promises.add(promise);
+            delegate.resolveAll(q, EMPTY_ADDITIONALS, promise);
+            promise.addListener(listener);
+        });
+        configureTimeout(questions, logPrefix, aggregatedPromise, promises.build());
         return aggregatedPromise;
     }
 
+    private void configureTimeout(List<DnsQuestion> questions, String logPrefix,
+                                  Promise<List<DnsRecord>> result,
+                                  List<Promise<List<DnsRecord>>> promises) {
+        if (queryTimeoutMillis == 0) {
+            return;
+        }
+        eventLoop.schedule(() -> {
+            if (result.isDone()) {
+                // Received a response before the query times out.
+                return;
+            }
+            final DnsTimeoutException exception = new DnsTimeoutException(
+                    '[' + logPrefix + "] " + questions + " are timed out after " +
+                    queryTimeoutMillis + " milliseconds.");
+            result.tryFailure(exception);
```

Comment:
setFailure? I think we're not checking the return value since we don't expect the result to be completable since we own the event loop so may as well make that explicit.

Reply:
Yeah, that makes sense. Fixed.